Fix loadBefore vs GC

navytux opened the pull request:
TemporaryStorage uses

    ._index   {} oid -> serial
    ._opickle {} oid -> data

as data store for load. However for loadBefore

    ._conflict_cache  {} (oid,serial) -> (data, time)

is reused as data store.

That would be ok if that place would be treated as data store, but given
its primary purpose - as its name suggests - was originally to be a
cache to resolve conflicts, it is "logical" that entries in this cache
are garbage-collected when entry age becomes > gc threshold.

Only now there is a problem: if an object is committed once, and time
passes, corresponding entry in ._conflict_cache will be removed. And this
would manifest itself as

    - load(oid)  -> gives latest data for the object (obtained via ._index and ._opickle)

but

    - loadBefore(oid, @head)  -> gives POSKeyError

-> Fix it by always preserving latest object revision in
._conflict_cache from being removed on GC. 

The fix is important for systems that use ZODB5, or ZODB4-wc2[1] because
there ZODB.Connection switched from primarily using load to exclusively
using loadBefore.

/cc @icemac @mauritsvanrees @mgedmin @d-maurer @dwt @hannosch

Fixes: https://github.com/zopefoundation/tempstorage/issues/8
[1] https://lab.nexedi.com/nexedi/ZODB/commit/8e7eab33

mauritsvanrees reviewed (approved):
Seems okay, but I am not familiar with the code at this level.

navytux commented:
Thanks, @mauritsvanrees. Let's get CI fixed first (https://github.com/zopefoundation/tempstorage/pull/17). And for the changes here I plan to post an amendment today.

d-maurer commented on src/tempstorage/TemporaryStorage.py:
```diff
@@ -124,11 +124,20 @@ def getSize(self):
+
+            # gc entries but keep latest record for each oid
+            for oid, hist in byoid.items():
+                hist.sort(key=lambda _: _[0]) # by serial
```
You could use `heapq.nlargest` here (as your only interest is to determine the highest serial) -- not sure whether it is important, though.

navytux replied:
Thanks. I would prefer to keep this code in the most simplest form for now. If we really see that this place starts to show in profiles or something like that, then it would be better to redo TemporaryStorage data structures and just keep all oid entries in rev↑ order in some single place, instead of `._index` + `.__conflict_cache`.

d-maurer reviewed (approved):
Please document your fix in `CHANGES.rst`.

Older `Zope` versions defined a `LowConflictConnection` class (in `Products.TemporaryFolder`) for use with sessions: it did not raise `ReadConflictError` (by handling invalidation only at transaction boundaries) and always read the most recent object revision: the latter part would mean transaction isolation "read committed". As `tempstorage` maintains few history, it might be interesting to get such a functionality again.

navytux commented:
The change was amended:

- rebased on top of master to get CI working;
- amended tests to actually exercise `loadBefore` in addition to `_conflict_cache`. This way we make sure that the fix actually affect its target - `loadBefore` (see 3d09f75e);
- added changelog entry + undeprecated the package (see a64adbac).

@mauritsvanrees, @dataflake, @d-maurer, thanks, once again, for your approvals.
Does updated change still look ok?

> Older `Zope` versions defined a `LowConflictConnection` class (in `Products.TemporaryFolder`) for use with sessions: it did not raise `ReadConflictError` (by handling invalidation only at transaction boundaries) and always read the most recent object revision: the latter part would mean transaction isolation "read committed". As `tempstorage` maintains few history, it might be interesting to get such a functionality again.

In ZODB5 `ReadConflictError` on non-transaction boundaries are generally no longer possible(*), because those read conflicts are gone since Zope 2.8 - see e.g. https://github.com/zopefoundation/ZODB/commit/3a493b01 for details.

Regarding "read committed" isolation level, maybe the right place to do it is at `ZODB.Connection` (or `ZODB.DB`) level. If an application exposes such desire to weaken isolation property, and at which level, and potentially with different levels depending on particular storage in multi-db setups, at DB creation time, then it would be just simple amendment to `ZODB.Connection` code to take that into effect, and instead of calling `loadBefore(obj, zconn.before)` to load an object, call `loadBefore(obj, maxtid)`.

For sessions and TemporaryStorage use-case it would be to set iso="read-committed" for `/temp_folder` mountpoint. But being able to adjust isolation level is generally useful not only for sessions. That's why I suggest to make it a generic place.

(*) except read-conflict wrt concurrent pack.

dataflake commented:
> Does updated change still look ok

Looks OK to me.

navytux commented:
Thanks, @dataflake. Let's merge it then. I hope it is ok, as the amendments were only about extending tests and updating changes/readme.

navytux commented:
What is the best way forward to make tempstorage 5.2 egg released?

dataflake commented:
I can take that on, I want to do some packaging cleanups first, though.

navytux commented:
Thanks, @dataflake.
